streams.go

```diff
@@ -0,0 +1,296 @@
+package lksdk
+
+import (
+	"sync"
+
+	protocol "github.com/livekit/protocol/livekit"
+)
+
+type baseStreamInfo struct {
+	Id         string
+	MimeType   string
+	Topic      string
+	Timestamp  int64
+	Size       *uint64
+	Attributes map[string]string
+}
+
+type ByteStreamInfo struct {
+	*baseStreamInfo
+	Name *string
+}
+
+type TextStreamInfo struct {
+	*baseStreamInfo
+}
+
+const (
+	STREAM_CHUNK_SIZE = 15_000
+)
+
+type StreamTextOptions struct {
+	Topic                 string
+	DestinationIdentities []string
+	StreamId              *string
+	ReplyToStreamId       *string
+	TotalSize             *uint64
+	Attributes            map[string]string
+	OnProgress            *func(progress float64)
+}
+
+type StreamBytesOptions struct {
+	Topic                 string
+	MimeType              string
+	DestinationIdentities []string
+	StreamId              *string
+	TotalSize             *uint64
+	Attributes            map[string]string
+	OnProgress            *func(progress float64)
+	FileName              *string
+}
+
+type baseStreamWriter[T any] struct {
+	onWrite func(data T)
+	onClose func()
+}
+
+func newBaseStreamWriter[T any](onWrite func(data T), onClose func()) *baseStreamWriter[T] {
+	return &baseStreamWriter[T]{
+		onWrite: onWrite,
+		onClose: onClose,
+	}
+}
+
+func (w *baseStreamWriter[T]) Write(data T) {
+	w.onWrite(data)
+}
+
+func (w *baseStreamWriter[T]) Close() {
+	w.onClose()
+}
+
+func writeStreamBytes(data []byte, e *RTCEngine, streamId string, destinationIdentities []string, totalSize *uint64, onProgress *func(progress float64)) {
+	chunkIndex := uint64(0)
+	for i := 0; i < len(data); i += STREAM_CHUNK_SIZE {
+		end := i + STREAM_CHUNK_SIZE
+		if end > len(data) {
+			end = len(data)
+		}
+
+		e.waitForBufferStatusLow(protocol.DataPacket_RELIABLE)
+		e.publishStreamChunk(&protocol.DataStream_Chunk{
+			StreamId:   streamId,
+			Content:    data[i:end],
+			ChunkIndex: chunkIndex,
+		}, destinationIdentities)
+
+		if onProgress != nil && totalSize != nil {
+			progress := float64(i) / float64(*totalSize)
+			(*onProgress)(progress)
+		}
+
+		chunkIndex++
+	}
+}
+
+type TextStreamWriter struct {
+	*baseStreamWriter[string]
+	Info TextStreamInfo
+}
+
+func newTextStreamWriter(info TextStreamInfo, header *protocol.DataStream_Header, e *RTCEngine, destinationIdentities []string, onProgress *func(progress float64)) *TextStreamWriter {
+	e.publishStreamHeader(header, destinationIdentities)
+
+	onWrite := func(data string) {
+		writeStreamBytes([]byte(data), e, info.Id, destinationIdentities, info.Size, onProgress)
```

Comment:
this should have proper utf8 splitting so that we don't split a string at a multi-byte character

Reply:
there's an edge case where a utf-8 character (e.g. an emoji) is at the exact end of a given stream chunk. 

So this multi-byte character would get split in the middle. I think we still need to avoid that. See https://github.com/livekit/client-sdk-js/blob/main/src/room/utils.ts#L629-L652

Reply:
Interesting, let me check that up!

Reply:
This is done in 1723103e5e35964881d8c514d4d9fa9952c4f536, please have a look.

cc @boks1971